include/crocoddyl/multibody/contacts/contact-2d.hxx

```diff
@@ -0,0 +1,120 @@
+
+template <typename Scalar>
+ContactModel2DTpl<Scalar>::~ContactModel2DTpl() {}
+
+template <typename Scalar>
+void ContactModel2DTpl<Scalar>::calc(const boost::shared_ptr<ContactDataAbstract>& data,
+                                     const Eigen::Ref<const VectorXs>&) {
+  Data* d = static_cast<Data*>(data.get());
+  pinocchio::updateFramePlacement(*state_->get_pinocchio().get(), *d->pinocchio, xref_.frame);
+  d->v = pinocchio::getFrameVelocity(*state_->get_pinocchio().get(), *d->pinocchio, xref_.frame);
+  const Vector3s& vw = d->v.angular();
+  const Vector3s& vv = d->v.linear();
+
+  pinocchio::getFrameJacobian(*state_->get_pinocchio().get(), *d->pinocchio, xref_.frame, pinocchio::LOCAL, d->fJf);
+  d->Jc.row(0) = d->fJf.row(0);
+  d->Jc.row(1) = d->fJf.row(2);
+
+  d->a = pinocchio::getFrameAcceleration(*state_->get_pinocchio().get(), *d->pinocchio, xref_.frame);
+  d->a0[0] = d->a.linear()[0] + vw[1]*vv[2] - vw[2]*vv[1];
+  d->a0[1] = d->a.linear()[1] + vw[0]*vv[1] - vw[1]*vv[0];
+
+  if (gains_[0] != 0.) {
+    d->a0[0] += gains_[0] * (d->pinocchio->oMf[xref_.frame].translation() - xref_.oxf)[0];
+    d->a0[1] += gains_[0] * (d->pinocchio->oMf[xref_.frame].translation() - xref_.oxf)[2];
+  }
+  if (gains_[1] != 0.) {
+    d->a0[0] += gains_[1] * vv[0];
+    d->a0[1] += gains_[1] * vv[2];
+  }
+}
+
+template <typename Scalar>
+void ContactModel2DTpl<Scalar>::calcDiff(const boost::shared_ptr<ContactDataAbstract>& data,
+                                         const Eigen::Ref<const VectorXs>&) {
+  Data* d = static_cast<Data*>(data.get());
+  pinocchio::getJointAccelerationDerivatives(*state_->get_pinocchio().get(), *d->pinocchio, d->joint, pinocchio::LOCAL,
+                                             d->v_partial_dq, d->a_partial_dq, d->a_partial_dv, d->a_partial_da);
+  const std::size_t& nv = state_->get_nv();
+  pinocchio::skew(d->vv, d->vv_skew);
+  pinocchio::skew(d->vw, d->vw_skew);
+  d->fXjdv_dq.noalias() = d->fXj * d->v_partial_dq;
+  d->fXjda_dq.noalias() = d->fXj * d->a_partial_dq;
+  d->fXjda_dv.noalias() = d->fXj * d->a_partial_dv;
+      
+  d->da0_dx.leftCols(nv).row(0).noalias() = d->fXjda_dq.row(0) +
+      d->vw_skew.row(0) * d->fXjdv_dq.template topRows<3>() -
+      d->vv_skew.row(0) * d->fXjdv_dq.template bottomRows<3>();
+  d->da0_dx.leftCols(nv).row(1).noalias() = d->fXjda_dq.row(2) +
+      d->vw_skew.row(2) * d->fXjdv_dq.template topRows<3>() -
+      d->vv_skew.row(2) * d->fXjdv_dq.template bottomRows<3>();
+
+  d->da0_dx.rightCols(nv).row(0).noalias() = d->fXjda_dv.row(0) + d->vw_skew.row(0) * d->fJf.template topRows<3>() - d->vv_skew.row(0) * d->fJf.template bottomRows<3>();
+  d->da0_dx.rightCols(nv).row(1).noalias() = d->fXjda_dv.row(2) + d->vw_skew.row(2) * d->fJf.template topRows<3>() - d->vv_skew.row(2) * d->fJf.template bottomRows<3>();
+  
+  if (gains_[0] != 0.) {
+    d->oRf = d->pinocchio->oMf[xref_.frame].rotation();
+    MatrixXs oRf2D(2,2);
```

Comment:
I haven't looked at how to optimize this part yet. However, for now, you could use `Matrix2s` instead of MatrixXs. Always prefer fixed-size matrices if possible.

Reply:
Did you forgot to change this?

Reply:
When I try to use Matrix2s (`typedef typename MathBase::Matrix2s Matrix2s;` in .hpp), I get the error `no type named 'Matrix2s' in 'crocoddyl::MathBaseTpl<double>`